pyteal/ast/frame.py

```diff
@@ -0,0 +1,188 @@
+from typing import TYPE_CHECKING
+
+from pyteal.ast.expr import Expr
+from pyteal.types import TealType, require_type
+from pyteal.errors import TealInputError, verifyProgramVersion
+from pyteal.ir import TealBlock, TealSimpleBlock, TealOp, Op
+
+if TYPE_CHECKING:
+    from pyteal.compiler import CompileOptions
+
+
+class Proto(Expr):
+    def __init__(self, arg_num: int, ret_num: int):
+        super().__init__()
+        if arg_num < 0:
+            raise TealInputError(f"subroutine arg number {arg_num} must be >= 0")
+        if ret_num < 0:
+            raise TealInputError(f"return value number {ret_num} must be >= 0")
+        self.arg_num = arg_num
+        self.ret_num = ret_num
```

Comment:
```suggestion
    def __init__(self, num_args: int, num_returns: int):
        super().__init__()
        if num_args < 0:
            raise TealInputError(f"the number of arguments provided to Proto must be >= 0 but {num_args=}")
        if num_returns < 0:
            raise TealInputError(f"the number of return values provided to Proto must be >= 0 but {num_returns=}")
        self.num_args = num_args
        self.num_returns = num_returns
```

one possible improvement - pls. take with a grain of salt

Reply:
looks right, I am taking